File: backend/routers/categories.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
import uuid
from database import get_db
from models import Category
from schemas import CategoryCreate, CategoryUpdate, CategoryRead

router = APIRouter()


@router.get("", response_model=List[CategoryRead])
def list_categories(db: Session = Depends(get_db)):
    return db.query(Category).order_by(Category.name).all()


@router.post("", response_model=CategoryRead)
def create_category(data: CategoryCreate, db: Session = Depends(get_db)):
    category = Category(
        id=str(uuid.uuid4()),
        name=data.name,
        icon=data.icon,
        color=data.color,
        type=data.type,
    )
    db.add(category)
    db.commit()
    db.refresh(category)
    return category


@router.get("/{category_id}", response_model=CategoryRead)
def get_category(category_id: str, db: Session = Depends(get_db)):
    category = db.query(Category).filter(Category.id == category_id).first()
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    return category


@router.put("/{category_id}", response_model=CategoryRead)
def update_category(category_id: str, data: CategoryUpdate, db: Session = Depends(get_db)):
    category = db.query(Category).filter(Category.id == category_id).first()
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    for field, value in data.model_dump(exclude_none=True).items():
        setattr(category, field, value)
    db.commit()
    db.refresh(category)
    return category


@router.delete("/{category_id}")
def delete_category(category_id: str, db: Session = Depends(get_db)):
    category = db.query(Category).filter(Category.id == category_id).first()
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    db.delete(category)
    db.commit()
    return {"ok": True}

File: src/api/categories.ts
import api from './client';
import { Category } from '../types';

export interface CategoryPayload {
  name: string;
  icon: string;
  color: string;
  type: string;
}

export const categoriesApi = {
  list: () => api.get<Category[]>('/categories').then(r => r.data),
  get: (id: string) => api.get<Category>(`/categories/${id}`).then(r => r.data),
  create: (data: CategoryPayload) => api.post<Category>('/categories', data).then(r => r.data),
  update: (id: string, data: Partial<CategoryPayload>) =>
    api.put<Category>(`/categories/${id}`, data).then(r => r.data),
  delete: (id: string) => api.delete(`/categories/${id}`).then(r => r.data),
};
